### PR TITLE
Makes the GUI talk to the controller over the serial port

### LINKED ISSUE #4: GUI should exchange state with the controller using the communication protocol
**What has to be done**
The controller talks [the following communication protocol](https://github.com/RespiraWorks/VentilatorSoftware/blob/master/common/generated_libs/network_protocol/network_protocol.proto), using nanopb.
Essentially: GUI sends `GuiStatus` containing desired mode parameters, Controller sends `ControllerStatus` containing sensor readings and other controller-side info.

As of writing, the GUI is stubbed out and not connected to the controller in any way.

Success criterion for this issue is that:
- We have a C++ abstraction somewhere in GUI code encapsulating talking over the protocol, for example: `ControllerStatus exchangeStatus(GuiStatus status)` (symmetric with the respective [controller-side code](https://github.com/RespiraWorks/VentilatorSoftware/blob/9309603783f6342876b0c9bab63bc43bfa2ce64c/controller/src/main.cpp#L66)) - this is a reasonable independent sub-task of this bug. @Miceuz was planning to do this.
- The GUI has a periodic thread that frequently calls that function, giving it parameters somehow obtained from UI element state and storing the resulting ControllerStatus'es in a form from which the UI can be rendered. @jkff is planning to do this.

**How do you know it has to be done**
That the GUI needs to talk to the controller is self-evident.
The communication protocol is settled, code linked above.
As for whether the exchange should happen periodically or via some other mechanism - I don't know, periodically seems reasonable.

A point of ambiguity is, exactly how is the GUI connected to the controller - I (@jkff) was assuming it will be over serial port but it should be clarified, as I know nothing about serial ports.

**Code pointers**
I (@jkff) think the whole thing should look roughly like this:
```
interface ConnectedDevice {
	ControllerState exchangeState(GuiState state)  // nanopb protos
}
main() {
	ConnectedDevice device = ConnectedDeviceOverSerialPort(...serial port parameters...);  // but initially a fake
	History<ControllerState> state_history(30 seconds);  // just a vector that pushes out too old stuff
	GuiState desired_state;
	on timer separate from QT {
		history.Add(device.ExchangeState(desired_state))
	}
	...bind desired_state to parameters set in the UI...
	...bind state_history to graphs and alarms displayed in the UI...
}
```

The current bug is about doing the ConnectedDevice stuff.

### PR DESCRIPTION
**Pull request by Miceuz**
Implement serial communication from GUI side. 
Fixes #4

### CODE REVIEW

**Comment by Miceuz** [comment]
Should we return bool from ReceiveControllerStatus, SendGuiStatus to report errors upstream?

File: gui/RespiraConnectedDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      QByteArray responseData = m_serial->readAll();
+
+      // continue reading characters until we read all we expect or we don't see
+      // any data for long enough time to estimate failure
+      while (responseData.length() < ControllerStatus_size &&
+             m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        responseData += m_serial->readAll();
+      }
+
+      if (ControllerStatus_size == responseData.length()) {
```
Comment (jlebar) [inline gui/RespiraConnectedDevice.cpp:62]: I don't think this is right.  The proto messages are of variable length, and will usually be much shorter than ControlleStatus_size.  `ControllerStatus_size` is the max length.

File: gui/RespiraConnectedDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
```
[the remark this answers is not in the document]
Comment (jkff) [reply gui/RespiraConnectedDevice.cpp:10]: The GUI uses stuff from `chrono.h`, which has a `DurationMs`. Maybe we can merge the two at some point, but for now I'd suggest to maintain consistency within a component. Agreed about constants though.

File: gui/RespiraConnectedDevice.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+    if (m_serial->bytesAvailable() > 0) {
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        m_serial->clear(QSerialPort::Input);
+      }
+    }
+
+    if (m_serial->waitForReadyRead(m_interframeTimeout_ms)) {
+      QByteArray responseData = m_serial->readAll();
```
Comment (jkff) [inline gui/RespiraConnectedDevice.cpp:53]: Can we preallocate this with `.reserve()`? I guess ControllerStatus might be small enough that it doesn't matter though.

File: gui/RespiraConnectedDevice.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  }
+
+  void ReceiveControllerStatus(ControllerStatus *controller_status) override {
+    // Drop everything we have in buffer right now as we are not sure if this is
+    // a full frame or a part of it.
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+    if (m_serial->bytesAvailable() > 0) {
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
```
Comment (jkff) [inline gui/RespiraConnectedDevice.cpp:47]: I feel like this should have some limit on number of iterations and return an error if it's too many.

File: gui/RespiraConnectedDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+      if (ControllerStatus_size == responseData.length()) {
+        pb_istream_t stream = pb_istream_from_buffer(
+            (const uint8_t *)responseData.data(), responseData.length());
+        if (!pb_decode(&stream, ControllerStatus_fields, controller_status)) {
+          // TODO: Log an error.
+        }
+      } else {
+        // TODO we have read some garbage here. Log an error. Raise an alert
+        return;
```
Comment (jkff) [inline gui/RespiraConnectedDevice.cpp:70]: nit: instead of a bunch of conditions "if (blah) { foo } else {return}", I'd suggest to invert and get: if (!blah) return; foo - this way foo has reduced nesting which is easier to read.

File: gui/RespiraConnectedDevice.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
+      : m_serial(new QSerialPort()), m_writeTimeout_ms(write_timeout_ms),
```
Comment (jkff) [inline gui/RespiraConnectedDevice.cpp:11]: When is the QSerialPort deleted?

File: gui/ProjectVentilatorGUI.pro [hunk 7]
```diff
@@ -1,14 +1,25 @@
-QT += core quick charts
+QT += core quick charts serialport
 CONFIG += c++17
 DEFINES += QT_DEPRECATED_WARNINGS
 
-SOURCES += $$files("*.cpp")
+SOURCES += $$files("*.cpp") \
+    ../common/generated_libs/network_protocol/network_protocol.pb.c \
```
Comment (jkff) [inline gui/ProjectVentilatorGUI.pro:6]: Is this necessary? Line 6 (now 10) is supposed to take care of this.

File: gui/ProjectVentilatorGUI.pro [hunk 8]
```diff
@@ -1,14 +1,25 @@
...
+    ../common/third_party/nanopb/pb.h \
+    ../common/third_party/nanopb/pb_common.h \
+    ../common/third_party/nanopb/pb_decode.h \
+    ../common/third_party/nanopb/pb_encode.h
 HEADERS += $$files("$$PWD/../common/**/*.h")
 INCLUDEPATH += $$PWD/../common/third_party/nanopb
 RESOURCES += qml.qrc Logo.png
-DISTFILES += Logo.png
+DISTFILES += Logo.png \
+    ../common/generated_libs/network_protocol/network_protocol.proto \
```
Comment (jkff) [inline gui/ProjectVentilatorGUI.pro:21]: I don't think these files have to be in DISTFILES - they aren't involved in building the GUI at all.

File: gui/RespiraConnectedDevice.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
+#include "../common/generated_libs/network_protocol/network_protocol.pb.h"
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+#include <QThread>
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:7]: This include is not necessary anymore

File: gui/RespiraConnectedDevice.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+#define INTER_FRAME_TIMEOUT 11
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+public:
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:14]: Per previous comments, it's better to hardcode these constants here - main.cpp doesn't know their values any better than the current file does.

File: gui/RespiraConnectedDevice.h [hunk 11]
```diff
@@ -0,0 +1,115 @@
+#include "../common/generated_libs/network_protocol/network_protocol.pb.h"
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+#include <QThread>
+
+#define INTER_FRAME_TIMEOUT 11
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:9]: Please use "inline constexpr int INTERFRAME_TIMEOUT_MS" or something (i.e. constant is better than a define; and it's good to include units)

File: gui/RespiraConnectedDevice.h [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+public:
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
+      : serialPortName(portName), m_writeTimeout_ms(write_timeout_ms),
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h]: Per previous comments, please name the private variables as some_private_var_ rather than m_privateVar

File: gui/RespiraConnectedDevice.h [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        m_interframeTimeout_ms(interframe_timeout_ms) {}
+
+  ~RespiraConnectedDevice() {
+    if (m_serial != nullptr) {
+      m_serial->close();
+    }
+  }
+  void createPort() {
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h]: This should return a bool too, so that Send and Receive can return an error if they can't even open the port

File: gui/RespiraConnectedDevice.h [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    if (m_serial != nullptr)
+      return;
+    printf("Socket create\n");
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:27]: Please remove unnecessary debug logging throughout the PR (except for that which you think is necessary)

File: gui/RespiraConnectedDevice.h [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    m_serial = new QSerialPort();
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h]: Use an std::unique_ptr for this variable and say `m_serial = std::make_unique<QSerialPort>()`, so that it is deleted in the destructor rather than memory-leaking

File: gui/RespiraConnectedDevice.h [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    m_serial->setPortName(serialPortName);
+    m_serial->setBaudRate(QSerialPort::Baud115200);
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h]: Please add comments explaining why these specific parameter values were chosen. Are all of them necessary btw?

File: gui/RespiraConnectedDevice.h [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    m_serial->setDataBits(QSerialPort::Data8);
+    m_serial->setParity(QSerialPort::NoParity);
+    m_serial->setStopBits(QSerialPort::OneStop);
+    m_serial->setFlowControl(QSerialPort::NoFlowControl);
+    if (!m_serial->open(QIODevice::ReadWrite)) {
+      printf("Could not open serial port %s\n",
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:36]: In QT, you can include `<QtDebug>` and log like this: `qCritical() << "Could not open serial port " << serialPortName;`
There's also qDebug, qInfo, qFatal (crashes if anything is logged to it).

File: gui/RespiraConnectedDevice.h [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+#include <QThread>
+
+#define INTER_FRAME_TIMEOUT 11
+
+class RespiraConnectedDevice : public ConnectedDevice {
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:11]: Please document what this does and mention the thread safety considerations: specifically, for the moment, Send and Receive have to be called from the same thread.

File: gui/RespiraConnectedDevice.h [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (!m_serial->waitForBytesWritten(m_writeTimeout_ms)) {
+      // TODO communication failure, port closed? Log an error and raise
+      // an alert
+      return false;
+    }
+    return true;
+  }
+
+#include <chrono>
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:60]: Please include "chrono.h" instead, then you don't need to repeat the `using`; and the include should be at the top of the file.

File: gui/RespiraConnectedDevice.h [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  using SteadyClock = std::chrono::steady_clock;
+
+  bool ReceiveControllerStatus(ControllerStatus *controller_status) override {
+    createPort();
+    // Drop everything we have in buffer right now as we are not sure if this is
+    // a full frame or a part of it.
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+
+    //    if (m_serial->bytesAvailable() > 0) {
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h]: Remove commented-out code

File: gui/RespiraConnectedDevice.h [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    if (m_serial->waitForReadyRead(m_interframeTimeout_ms)) {
+      QByteArray responseData = m_serial->readAll();
+
+      // continue reading characters until we don't see
+      // any data for long enough time to estimate end of packet silence
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        responseData += m_serial->readAll();
+      }
+
+        pb_istream_t stream = pb_istream_from_buffer(
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:90]: This doesn't look clang-formatted, please make sure you have the pre-commit hooks installed.

File: gui/RespiraConnectedDevice.h [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+            (const uint8_t *)responseData.data(), responseData.length());
+
+        if (!pb_decode(&stream, ControllerStatus_fields, controller_status)) {
+          printf("ERROR: Could not decode ControllerStatus\n");
```
Comment (jkff) [inline gui/RespiraConnectedDevice.h:94]: return false?

File: utils/decoder.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
+import serial
```
Comment (jkff) [inline utils/decoder.py:1]: I would suggest to put the python scripts in a separate PR, if you think they should be checked in at all.

File: gui/main.cpp [hunk 24]
```diff
@@ -37,39 +38,24 @@ int main(int argc, char *argv[]) {
...
-            // Fill the status with fake data.
-            controller_status->uptime_ms =
-                TimeAMinusB(SteadyClock::now(), startup_time).count();
-            auto *sensors = &controller_status->sensor_readings;
-            sensors->pressure_cm_h2o =
-                sin(controller_status->uptime_ms * 0.001);
-            sensors->volume_ml = 0;
-            sensors->flow_ml_per_min = 0;
-          });
+      std::make_unique<RespiraConnectedDevice>("/dev/pts/6", 10,
```
Comment (jkff) [inline gui/main.cpp:41]: Please keep the old FakeConnectedDevice here by default (so that a contributor who launches the GUI without the scripts or a real device at least sees some fake data), and put the /dev/pts code commented-out instead, and add a TODO to make it configurable using a command-line flag.

File: gui/RespiraConnectedDevice.cpp (hunk 6 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.cpp]: destructor added to close it.

File: gui/RespiraConnectedDevice.cpp (hunk 4 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.cpp]: no, waitForReadyRead takes care of that - it timeouts and returns false.

File: gui/RespiraConnectedDevice.cpp (hunk 1 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.cpp]: Yeah, that wasn't right. Didn't realize we have variable length alerts list

File: gui/RespiraConnectedDevice.cpp [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    m_serial->write((const char *)tx_buffer, stream.bytes_written);
+
+    if (!m_serial->waitForBytesWritten(m_writeTimeout_ms)) {
+      // TODO communication failure, port closed? Log an error and raise an
+      // alert
+      return;
+    }
+  }
+
+  void ReceiveControllerStatus(ControllerStatus *controller_status) override {
```
[the remark this answers is not in the document]
Comment (Miceuz) [reply gui/RespiraConnectedDevice.cpp]: Due to threading issues now SendGuiStatus is always called after ReceiveControllerStatus. So we basically become synchronous where Cycle Controller sets the pace. I was thinking about this a bit last week, could not think of pros to having async exchange while cons of sync is obvious.

File: utils/decoder.py (hunk 23 above)
Comment (jlebar) [reply utils/decoder.py:1]: Having them checked in might be nice!  It would obviate the need for the `DEV_MODE_comms_handler`.

Alternately we handle everything in the GUI.

**Comment by jlebar** [comment]
I defer to @jkff's review, don't wait for my review.

File: gui/RespiraConnectedDevice.cpp (hunk 6 above)
Comment (martukas) [reply gui/RespiraConnectedDevice.cpp]: There might be a more Qt-idiomatic way to deal with cleanup if ConnectedDevice inherited from QObject, as already does QSerialPort. The parent destroys all children, as described [here](https://doc.qt.io/qt-5/qobject.html#QObject). Likewise, it seems PeriodicClosure might exist as a [QThread](https://doc.qt.io/qt-5/qthread.html), which would give you the signals/slots mechanism (which you may or may not need?).  I also see the merits of using the possibly more familiar std machinery over Qt's. Upon further examination of the code, I suspect this is irrelevant to how we intend this to work. Just wanted to make you aware of these tools in case Qt is less familiar to you guys.

File: gui/RespiraConnectedDevice.h (hunk 14 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.h]: yep!

File: gui/RespiraConnectedDevice.h (hunk 16 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.h]: These parameters fully specify the serial port configuration we use, some of them are default (like parity, data bits, stop bits) but I think it's good to be explicit here.

File: gui/RespiraConnectedDevice.h (hunk 18 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.h:11]: Done

File: gui/RespiraConnectedDevice.h (hunk 20 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.h]: Yep. I still don't like the fact that we potentially try to deserialize some garbage, but having this here and now creates too much packet timeouts. This will be non-issue when framing comes along.

File: gui/RespiraConnectedDevice.h (hunk 22 above)
Comment (Miceuz) [reply gui/RespiraConnectedDevice.h]: yep

File: gui/main.cpp (hunk 24 above)
Comment (Miceuz) [reply gui/main.cpp:39]: Done
